File: 42. Trapping Rain Water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # edge case
        if len(height) < 3:
            return 0
        
        # get water stored above a column by looking left only
        left = [0]
        maxh = height[0]
        for h in height[1:]:
            if h < maxh:
                left.append(maxh - h)
            else:
                maxh = h
                left.append(0)
            
        # get water stored above a column by looking right only
        right = [0]
        maxh = height[-1]
        # height.reverse()
        for h in height[-2::-1]:
            if h < maxh:
                right.append(maxh - h)
            else:
                maxh = h
                right.append(0)
        right.reverse()
        
        # minimize the water content based on left and right view
        vol = 0
        for l, r in zip(left, right):
            vol += min(l, r)
            
        return vol
                
        

File: 42. Trapping Rain Water/solution.java
class Solution {
    public int trap(int[] height) {
        // edge case
        int vol = 0;
        if (height.length <= 2) {
            return vol;
        }
        
        int[] left = new int[height.length];
        int[] right = new int[height.length];
        int maxHeight = height[0];
        
        // look from left to right
        left[0] = 0;
        for(int i = 1; i < height.length; i++) {
            if (height[i] > maxHeight) {
                maxHeight = height[i];
                left[i] = 0;
            } else {
                left[i] = maxHeight - height[i];
            }
        }
        
        maxHeight = height[height.length-1];
        
        // look from left to right
        right[height.length-1] = 0;
        for(int i = height.length-2; i >= 0; i--) {
            if (height[i] > maxHeight) {
                maxHeight = height[i];
                right[i] = 0;
            } else {
                right[i] = maxHeight - height[i];
            }
        }
            
        // take the min of both views
        for(int i = 0; i < height.length; i++) {
            vol += Math.min(left[i], right[i]);
        }
        
        return vol;
    }
}
